### PR TITLE
Fix error handling in CMS_EncryptedData_encrypt

### LINKED ISSUE #21985: Memory leaks in CMS_EncryptedData_encrypt
```
../../util/shlib_wrap.sh ../../apps/openssl cms -EncryptedData_encrypt -in ../smcont.txt -outform PEM -aes128 -secretkey 000102030405060708090A0B0C0D0E0F -stream -out test.cms -noout

=================================================================
==5858==ERROR: LeakSanitizer: detected memory leaks

Direct leak of 16 byte(s) in 1 object(s) allocated from:
    #0 0x7f82d1bab09f in __interceptor_malloc ../../../../gcc-trunk/libsanitizer/asan/asan_malloc_linux.cpp:69
    #1 0x7f82d13f2be4 in ossl_cms_EncryptedContent_init crypto/cms/cms_enc.c:213
    #2 0x7f82d140926e in CMS_EncryptedData_encrypt_ex crypto/cms/cms_smime.c:238
    #3 0x443891 in cms_main apps/cms.c:1019
    #4 0x467522 in do_cmd apps/openssl.c:418
    #5 0x42611e in main apps/openssl.c:298
    #6 0x7f82d0939f44 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x21f44)

SUMMARY: AddressSanitizer: 16 byte(s) leaked in 1 allocation(s).
ed@w-ed:~/ossl/openssl/test/testutil$ ../../util/shlib_wrap.sh ../../apps/openssl cms -EncryptedData_encrypt -in ../smcont.txt -outform PEM -aes128 -stream -out test.cms -noout
Error creating CMS structure
C0E72ED8C37F0000:error:17000082:CMS routines:CMS_EncryptedData_set1_key:no key:crypto/cms/cms_enc.c:231:

=================================================================
==5864==ERROR: LeakSanitizer: detected memory leaks

Direct leak of 32 byte(s) in 1 object(s) allocated from:
    #0 0x7fc3d7cbb09f in __interceptor_malloc ../../../../gcc-trunk/libsanitizer/asan/asan_malloc_linux.cpp:69
    #1 0x7fc3d76e17b0 in CRYPTO_zalloc crypto/mem.c:197
    #2 0x7fc3d746cba3 in asn1_item_embed_new crypto/asn1/tasn_new.c:136
    #3 0x7fc3d746cdbc in ASN1_item_new_ex crypto/asn1/tasn_new.c:41
    #4 0x7fc3d750c8a9 in CMS_ContentInfo_new_ex crypto/cms/cms_lib.c:54
    #5 0x7fc3d7519255 in CMS_EncryptedData_encrypt_ex crypto/cms/cms_smime.c:235
    #6 0x443891 in cms_main apps/cms.c:1019
    #7 0x467522 in do_cmd apps/openssl.c:418
    #8 0x42611e in main apps/openssl.c:298
    #9 0x7fc3d6a49f44 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x21f44)

SUMMARY: AddressSanitizer: 32 byte(s) leaked in 1 allocation(s).
```

### PR DESCRIPTION
**Pull request by bernd-edlinger**
That caused several memory leaks in case of error. Also when the CMS object that is created by CMS_EncryptedData_encrypt is not used in the normal way, but instead just deleted by CMS_ContentInfo_free some memory was lost.

Fixes #21985

##### Checklist

- [ ] documentation is added or updated
- [x] tests are added or updated

### CODE REVIEW

**Review by t8m — APPROVED** [review]
@bernd-edlinger If you promise making patches against newer branches I am OK merging this to 1.1.1 as this looks like a potential security hardening.

**Comment by bernd-edlinger** [comment]
OK, deal :)

**Comment by t8m** [comment]
We need a review from committers ASAP if this should go in before the 1.1.1 EOL

**Review by hlandau — APPROVED** [review]

**Review by beldmit — APPROVED** [review]
LGTM

File: crypto/cms/cms_sd.c
Comment (bernd-edlinger) [inline crypto/cms/cms_sd.c:755]: While working out a PR for master, I realised that this function
can be called twice if e.g. CMS_verify is called twice, and that would
create a memory leak.

**Comment by bernd-edlinger** [comment]
@t8m @hlandau @beldmit please re-confirm.

**Comment by beldmit** [comment]
My approval persists.

**Review by t8m — APPROVED** [review]

**Comment by bernd-edlinger** [comment]
Sorry, I had to update the PR once more, I added two fuzz/corpora/cms files, one
directly from the failed CI workflow in master which turned out to directly affect this PR too.
I believe I have now fixed the issue correctly, and just-in-time for the upcoming 1.1.1 release party :-)
So please everyone re-onfirm your approvals.
Thanks!

**Comment by bernd-edlinger** [comment]
Please re-approve.
Thanks!

**Review by beldmit — APPROVED** [review]
LGTM

**Review by t8m — APPROVED** [review]

**Review by t8m — APPROVED** [review]

**Comment by t8m** [comment]
Merged to 1.1.1 branch. Thank you for your contribution.

**Pull request closed without merging**
